### PR TITLE
spectrum-scanner: Application for radio ED scanning

### PR DESCRIPTION
**Pull request by gebart**
This application is used to generate a 3D spectrum of the radio medium.

~~Best used in conjunction with the script `plot_spectrum.py` introduced in https://github.com/RIOT-OS/RIOT/pull/7139~~

A plotting script is integrated in this PR.

### CODE REVIEW

File: spectrum-scanner/Makefile [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Set the name of your application:
+APPLICATION = spectrum-scanner
+
+# If no BOARD is found in the environment, use this default:
+BOARD ?= native
```
Comment (OlegHahm) [inline spectrum-scanner/Makefile]: Isn't native a pointless default platform for this application?

Comment (gebart) [inline spectrum-scanner/Makefile]: Just copy pasta. So far the only platform which supports the netopt extension is the kw41z.

**Comment by gebart** [comment]
Significantly improved the quality of the produced image by computing the time average of multiple measurements on the device side before sending the values to the PC.

**Comment by gebart** [comment]
![https://github.com/gebart/RIOT-applications/blob/c4ecac6c5e8d4a1f2ad8310cd624e17a2c50b890/spectrum-scanner/tools/example.png]

File: spectrum-scanner/main.c [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                    ed_average[k][ch] += expf(level * ln10);
+                }
+            }
+            ++count;
+            thread_yield();
+        } while(xtimer_usec_from_ticks64(xtimer_now64()) < target);
+        for (unsigned int k = 0; k < netif_numof; ++k) {
+            printf("[%u, %10" PRIu32 ", %u] ", k, xtimer_usec_from_ticks(xtimer_now()), count);
+            for (unsigned int ch = IEEE802154_CHANNEL_MIN; ch <= IEEE802154_CHANNEL_MAX; ++ch) {
+                ed_average[k][ch] = logf(ed_average[k][ch] / count) / ln10;
```
Comment (gebart) [inline spectrum-scanner/main.c]: I think this expression can be rewritten to avoid all divisions

**Comment by gebart** [comment]
switched to using fmt instead of printf

**Comment by smlng** [comment]
all dependencies in RIOT are merged, @gebart you think this is ready then? Though, it obviously needs some squashing 😉

**Comment by gebart** [comment]
I think this is ready

**Comment by gebart** [comment]
squashed

File: spectrum-scanner/README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
+About
+=====
+
+This application is designed to run together with the script
+`tools/plot_spectrum.py` to monitor energy levels on all
```
Comment (smlng) [inline spectrum-scanner/README.md]: from script name? There is `plot_rssi.py`.

File: spectrum-scanner/tools/README.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
+# RIOT Spectrum Scanner Application
+
+## About
+
+This script can be used to plot the radio signal spectrum when a connected node
+is running the [spectrum-scanner application](https://github.com/RIOT-OS/applications/tree/master/spectrum-scanner)
```
Comment (smlng) [inline spectrum-scanner/tools/README.md]: Now that script is part of this repo and within the scanner-app directory, this section can be shortened a bit, e.g. removing the links?

File: spectrum-scanner/tools/README.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+application located in [RIOTs application repository](https://github.com/RIOT-OS/applications).
+This node scans over the available radio channels performing CCA measurements
+and outputting the measured ED level via a serial port. This output is then
+parsed by the `plot_spectrum.py` script included in this folder run on a host computer.
```
Comment (smlng) [inline spectrum-scanner/tools/README.md]: adapt name here and below -> `plot_rssi.py`

**Comment by smlng** [comment]
@Lotterleben: as already discussed offline, this could be interesting for your research/work - and maybe a teaser to get you back to RIOT, too 😉 

It requires latest RIOT master (as of Sep. 10th, 2017), as some dependencies were merged quite recently.

**Comment by smlng** [comment]
@gebart what's you ETA for this 😉 Besides my minor comments regarding typos (see above), I'd like to get this merge swiftly.

**Comment by gebart** [comment]
Ah, I forgot this was still open. I'll fix the typos now.

**Comment by smlng** [comment]
amend or squash directly, and I'll merge after short tests.

**Comment by gebart** [comment]
fixed typos, amended

**Comment by smlng** [comment]
works like charm. However, one question: How can I provoke some spikes, I tried to ping on `ff02::1` from another node, but nothing happened i.e., is visible on the scanner picture.

**Comment by smlng** [comment]
ah, or is it because I'm using board `samr21-xpro` -> is it suitable for this?

**Comment by gebart** [comment]
The samr21-xpro should work.

You could try streaming some Youtube or something over wifi to generate some noise in the 2.4 GHz band and see if that makes a difference. Also, take a look at the numbers being printed over the UART (it's in plain text) and see that you're not just seeing constant values, they should be at least fluctuating a little bit in any environment.

So far I think only the at86rf2xx and the kw41zrf (proposed in RIOT-OS/RIOT#7107) drivers support the necessary CCA and RSSI netopts. I have used the samr21-xpro and the frdm-kw41z to test this script, I think the example.png file is from a frdm-kw41z run, the samr21-xpro seem to give a little bit lower numbers in general for the RSSI value (at least in my tests in an office environment with some wifi networks).

**Comment by gebart** [comment]
Also, note that ping in RIOT is by default 5 bytes long data, which means that the packets are only around 1 ms long. The scanner is only running at an effective duty cycle of 1-2 %, because of the limitations of the implementation. Right now everything is going through the kernel IPC to request a measurement followed by a request of the latest RSSI. The CCA checks would need to run continuously and stream RSSI measurements to provide a greater duty cycle.

**Comment by smlng** [comment]
@gebart thanks for the info, yeah WiFi from my phone brings some spikes. For the ping with RIOT: I used 100 bytes payload with very short interval, but still not much influence on the scanner. 

Anyway ACK and GO!

**Pull request merged**